code/lib/cli/src/check.ts

```diff
@@ -0,0 +1,118 @@
+ * This should reliably tell us if the user has `@storybook/core` in their project as a top-level dependency or not.
+ */
+function tryFindCore() {
+  try {
+    const found = require.resolve('@storybook/core');
+
+    if (!found.includes('node_modules')) {
+      // We're either in PNP-mode or linked-mode.
+      // We were able to find the core package, so let's hope for the best.
+      return true;
+    }
+
+    if (found.includes('npx')) {
+      // We're in npx-mode. Good to proceed.
+      return true;
+    }
+
+    const TopLevelNodeModulesPath = found.split('node_modules')[0];
+    const projectPackageJsonPath = findUp.sync('package.json', { cwd: TopLevelNodeModulesPath });
+
+    if (!projectPackageJsonPath) {
+      // Unknown how this could happen.. a users uses our package, and has `node_modules` in the path, but no `package.json`?
+      return false;
+    }
+
+    const projectPackageJsonContents = readJsonSync(projectPackageJsonPath);
+
+    const all = [
+      ...Object.keys(projectPackageJsonContents?.devDependencies || {}),
+      ...Object.keys(projectPackageJsonContents?.dependencies || {}),
+      ...Object.keys(projectPackageJsonContents?.peerDependencies || {}),
+    ];
+
+    if (all.includes('@storybook/core')) {
+      return true;
+    }
+
+    return false;
+  } catch (e) {
+    return false;
+  }
+}
+
+async function maybePrompt() {
+  if (process.env.CI) {
+    return false;
+  }
+
+  const { runMigration } = await prompts({
+    type: 'confirm',
+    initial: true,
+    name: 'runMigration',
+    message: dedent`
+      We detected you have not installed storybook core. Did you upgrade storybook recently without running the upgrade command?
+
+      At the very least we need to have the @storybook/core package installed in the project.
+      We recommend running the following command to upgrade storybook:
+      > npx storybook@latest upgrade
+
+      If that is for some reason not an option for you, follow this guide:
+      https://storybook.js.org/docs/react/workflows/migrating-to-8-2/
+
+      Would you want to run the migration now?
+      (answering no will exit the process)
+    `,
+  });
+
+  return runMigration;
+}
+
+export async function check() {
+  const packageJson = await readJSON(join(__dirname, '..', 'package.json'));
+
+  const core = tryFindCore();
+
+  if (core) {
+    return true;
+  }
+
+  const shouldRunMigration = await maybePrompt();
+
+  if (!shouldRunMigration) {
+    console.log(dedent`
+      You opted not to run the automatic migration. Please install @storybook/core manually.
+      
+      To understand this change, and why it was needed, read this guide:
+      https://storybook.js.org/docs/react/workflows/migrating-to-8-2/
+    `);
+    return false;
+  }
+
+  await execa('npx', [`-p`, `sb@${packageJson.version}`, 'sb', 'automigrate'], {
+    stdio: 'inherit',
+    env: {
+      CI: 'true',
+    },
+  });
```

Comment:
I don't trust execa to find `npx` without setting the `shell` setting to true. According to their latest docs, using the `shell` in execa should be avoided, though. Let's wrap this code in a try catch block and notify the user what to do if it fails.